### ISSUE TITLE
Reconsider handling of NaN and inf

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#2269]
Also see #2266.

This query just fails, with no obvious way to catch and handle the error:

```
r.expr([1e200]).map(r.expr(1e200).mul(r.row))

RqlRuntimeError: Non-finite number: inf in:
r([1e+200]).map(r(1e+200).mul(r.row))
```

That is bad, because it makes analytical queries fragile, and allows for an easy DoS attack if users can insert large floating point numbers into a table.

I personally would prefer having nan and inf as legal values in ReQL, though there is the question of how to encode them in JSON.

Another option would be to ignore these errors in operations such as `map` and `reduce`, though that seems like a bad idea.

Ideally we would have something like `default` but for illegal numbers, so users could handle these errors and have them not break their whole query.

**Comment by stuartpb** [comment]
What remains of this issue that wasn't sorted out with #1023?

**Comment by danielmewes** [comment]
@stuartpb all of it. #1023 just changed the arguments to `between`.
`inf` here refers to a special floating point number value that you get when you go out of the number range. The `r.maxval` from #1023 on the other hand is not a number, but an opaque typeless value larger than any key (including strings, arrays etc.).

**Comment by ajschlosser** [comment]
@danielmewes I have also run into this issue while using ```map``` to calculate data analytics before returning the results of a somewhat large query. In brief, I am grouping certain data by U.S. state; however, if one particular state does not have data for a particular range, I get the ```RqlRuntimeError``` for dividing by zero. Is there a more graceful way to handle this than conditional statements?

**Comment by coffeemug** [comment]
@ajschlosser -- unfortunately `default` doesn't work for the divide by zero error, so the best way to handle it is with conditionals. The error handling here is definitely an oversight on our part; we'll try to fix this as soon as we can.

**Comment by ajschlosser** [comment]
@coffeemug Thanks for the speedy response. I'll stick to conditionals for now, then. Looking forward to future updates!

**Comment by coffeemug** [comment]
Moving to subsequent, I think we should either add a `catch` term, or extend `default` to work on more error cases.

**Comment by mlucy** [comment]
I think we should handle this by extending `default` to handle `INF` and `NaN` errors.  The simplest thing to do would be to just say that we consider these errors non-existence errors, but that doesn't quite make sense so I would also be open to introducing a new error type `foo` and saying that default handles both non-existence errors and `foo` errors.

I'm not sure we should do this on the first pass, but I think `default` should be able to handle arbitrary errors.  We could do this with an optarg `handle` which is an array of strings specifying what error classes `default` should catch.  (In this case we should also make the function passed to `default` optionally take a second argument specifying the error type.)

**Comment by deontologician** [comment]
Are enough people running into this that there's a reason to do the quick fix and not just wait until we can make default work for everything? I think default with a function would be really nice

**Comment by mlucy** [comment]
@deontologician -- the question is what `default` should catch by default.  I think it should probably catch these `INF` and `NaN` errors by default, but probably shouldn't catch e.g. adding a string to a number.  So even if we make `default` work for everything we still need to decide whether to define these errors as non-existence errors to make `default` catch them by default, or whether to make a new error class and say that `default` also catches that error class by default.

**Comment by VeXocide** [comment]
In my opinion we shouldn't be extending the JSON format in such a way that it's no longer parseable by a strictly standards-compliant parser. I'd prefer to add a `catch` instead of overloading `default` to handle these cases as well as `default` does one thing at the moment and does it well.

**Comment by deontologician** [comment]
I like extending default vs. add a `catch` term because it keeps ReQL in the declarative realm. The idea is to force the query to return a value instead of erroring out. your client code should be catching exceptions.

Making `NaN` a non-existence error makes sense to me (your answer isn't a number, would you like to provide a default?). I think `Inf` is a borderline conceptually, but I think making a whole separate exception type is splitting hairs, so I vote for both becoming NonExistence errors.

**Comment by mlucy** [comment]
I like extending `default` more than adding `catch` just because we already have too many goddamn terms, and in general it seems unnecessarily complicated to have multiple constructs for error handling.

The question of whether `default` should handle `NaN` and `Inf` errors is less obvious to me, but it feels right.  I think of dividing by 0 as being in the same class of errors as reducing over an empty list or a variety of other edge cases that produce non-existence errors.

**Comment by VeXocide** [comment]
I don't view `default` as error handling, though my perception there may just be off.

**Comment by mlucy** [comment]
> Making NaN a non-existence error makes sense to me (your answer isn't a number, would you like to provide a default?). I think Inf is a borderline conceptually, but I think making a whole separate exception type is splitting hairs, so I vote for both becoming NonExistence errors.

I don't particularly object, so I guess the current state of the proposal for the first implementation is literally just "change several numeric errors to be non-existence errors".

**Comment by VeXocide** [comment]
I like the idea of changing `NaN` to become a `NonExistence` error, less sure about `Inf` though.

**Comment by danielmewes** [comment]
I don't think `NaN` is quite the same as non-existence. You can get `NaN` through perfectly fine calculations simply by encountering rounding issues.
I suggest introducing a `ReqlNumericError` type and making both `NaN` and `Inf` throw an instance of it.

Making `default` catch those by default feels slightly wrong to me. But it probably wouldn't be terrible.

**Comment by deontologician** [comment]
I guess I'm thinking of the very common case of division by zero. It would be really convenient to be able to provide a default for that. So I would be cool with division by zero errors being caught by `default` and the other numerical snafus which are harder to predict just being whatever exceptions they are now

**Comment by mlucy** [comment]
I think 99.99% of the time people encounter `NaN` or `Inf` in ReQL calculations, it won't be because they're doing some sort of complicated scientific computing task that occasionally generates `NaN` in the course of regular operation, it will be because they're doing bog-standard division and a 0 ended up in the denominator.  I think that you probably want `default` to handle that case.

I don't have a particularly strong opinion on making them non-existence eorrs vs. some new numeric type.

**Comment by danielmewes** [comment]
After giving this some more thought, I think I agree now that the division by 0 case is the most relevant one, and I think we should just make NaN a `NonExistence` error and make it catchable by `default`.
We should keep the handling of `Inf` the way it currently is.

I was originally concerned about DoS attacks by malicious users injecting values that are too large, and thereby breaking analytical queries that are using these values and could be made to overflow. However even if we provided a way to handle `Inf`, that wouldn't be too useful in most queries I can think of because one way or the other, the result of the analytical query would be either wrong or undefined. So that probably wouldn't even be very useful.

**Comment by danielmewes** [comment]
Long term I think NaNs should become their own exception type, and we should do what was suggested about making it possible to specify which error types `default` should handle.

**Comment by mlucy** [comment]
@danielmewes -- I'm a little confused by your comment.  Are you saying you think `NaN` should be a `NonExistence` error but `Inf` shouldn't be?  That seems really bad to me.

**Comment by danielmewes** [comment]
@mlucy Yes. I I think having `Inf` throw a non existence error would be extremely weird. I think it makes a more sense for `NaN`, because you can get it by dividing by `0`, which is often similar to not having a value. Like you said:
>  I think of dividing by 0 as being in the same class of errors as reducing over an empty list or a variety of other edge cases that produce non-existence errors.

This argumentation doesn't apply to `Inf` at all as far as I can tell.

**Comment by mlucy** [comment]
@danielmewes -- you usually get `Inf` when you divide by 0, unless you're doing `0/0`.

```rb
irb(main):225:0> 0.0/0.0
NaN
irb(main):226:0> 1.0/0.0
Infinity
```

**Comment by coffeemug** [comment]
> Making `default` catch those by default feels slightly wrong to me. But it probably wouldn't be terrible.

I think `default` should definitely catch both `NaN` and `Inf` by default. I don't think of `default` as something that catches non-existence errors (and I don't think our users do either). I think most people think of `default` as "there is some data problem in query, let's throw `default` in there to handle it". So `NaN` and `Inf` should definitely be handled.

I do think that making either `NaN` and `Inf` a subset of non-existence errors is strange. I'd much prefer having a separate error class for these, that `default` handles out of the box.

It would be nice to have an optarg to `default` to define which errors to handle (which effectively turns `default` into `catch`, which I think is what we want):

```py
r.expr(X).default(Y, catch_errors=['NonExitence', 'Math'])
```

I don't think the optarg is necessary for the first pass, though. I think relatively few people would actually use it, so I wouldn't bother with it until we get actual requests for this feature.

**Comment by danielmewes** [comment]
@mlucy I wasn't aware of that. That indeed makes it a bad idea to treat `NaN` and `Inf` differently.

I'd like to think about this for another day, but so far the following seems pretty reasonable to me then:
* adding a new error type (`ReqlNumericError`, `ReqlMathError`?)
* making `default` catch that error type in addition to non existence errors
* long-term (but not for 2.2): making `default` accept a list of error types to catch

(So basically what @coffeemug said.)

**Comment by mlucy** [comment]
Alright, that sounds good to me too.  (How about `ReqlFloatError`?)

**Comment by VeXocide** [comment]
Have a look at https://docs.python.org/3/library/exceptions.html for inspiration, it defines `ArithmeticError` as a base of `OverflowError`, `ZeroDivisionError`, and `FloatingPointError`.

**Comment by coffeemug** [comment]
> so far the following seems pretty reasonable to me then:
* adding a new error type (`ReqlNumericError`, `ReqlMathError`?)
* making `default` catch that error type in addition to non existence errors
* long-term (but not for 2.2): making `default` accept a list of error types to catch

:+1:

**Comment by danielmewes** [comment]
Actually another thing we could consider is to check for divisions by 0 explicitly in `div` and `mod`, and to delay general handling of `Inf` and `NaN` errors for later (maybe when we extend `default` to catch arbitrary error types).

This might make it easier to justify treating those errors as non-existence errors, which would reduce the amount of changes we would have to do to our code base and error hierarchy.
However I think doing this would still be a bit odd.

**Comment by danielmewes** [comment]
We would also need to do that anyway if we wanted to mimic the Python error types and have a `ZeroDivisionError`.

I would be happy to have `default` catch `ZeroDivisionError` (by default), while I still have some small doubts for having it handle all numerical errors.

So here's another proposal which is similar but slightly different:
* add a new error type `ReqlZeroDivisionError`
* make `default` catch that error type in addition to non existence errors
* make `div` and `mod` throw that error if one tries to divide by 0 (or sufficiently close to it to get an `Inf` result)
* long-term (but not for 2.2): making `default` accept a list of error types to catch. Also add a `ReqlNumericError` or similar to allow handling `Inf` and `NaN` in places other than division by zero

Any thoughts?

**Comment by coffeemug** [comment]
That sounds reasonable. The only difference is that we get to delay implementing general numeric errors, is that right?

**Comment by danielmewes** [comment]
Yes, and we delay the decision of whether general numeric errors should be caught by `default` by default.

**Comment by coffeemug** [comment]
That sounds good to me.

**Comment by mlucy** [comment]
I would prefer to just do this for all numeric errors.  It seems harder to mess and and miss something.

**Comment by ajschlosser** [comment]
A ``ReqlZeroDivisionError``caught by ``default`` would, I think, completely resolve the issue for me. We've encountered it several times in our project(s), and have had to come up with workarounds that might potentially become bottlenecks. I really appreciate the thought and attention that you all have given to this issue. Looking forward to how things proceed. :+1:

**Comment by deontologician** [comment]
I am leaning towards only `ZeroDivisionError` handled by `default` since it's the common case.

If I see a query like `r(1e200).div(x).default(0)`, do I expect this to handle precision related numeric errors?

Pros: People are likely to not think about edge cases in floating point math, so explicit errors would be nice. If you're testing, you'll get an informative blowup instead of having strangely incorrect results.

Cons: If you have a floating point edge case possible, and you don't think about it, your program could crash in production if we throw a special exception. These errors probably only happen rarely so unthorough testing might not catch it. We could "help" these applications not fail by swallowing these rarer errors in `default`, even though they were probably only expecting to catch zero division errors.

**Comment by coffeemug** [comment]
> People are likely to not think about edge cases in floating point math, so explicit errors would be nice.

In my experience that's not what people want for these queries. Most of the time they don't have control of the data (they get whatever's in the source), and when the query blows up they think "errr, why do I have to deal with this, I just want to easily set the default!"

I think the question of which numeric errors to support initially isn't that important since we'll eventually add a way to explicitly list which errors to catch, so I'd just do whatever's easiest from the development POV for now.

**Comment by mlucy** [comment]
Alright, my vote is to just give numeric errors a type and make `default` handle them.  It's easy to implement, we're unlikely to miss something, and it probably does what people want.

**Comment by deontologician** [comment]
ok, fair enough

**Comment by danielmewes** [comment]
Yeah I might have been over-thinking this.
I suspect if a user places a `default` in a given place, they usually intend to ignore any data-related issues in that part of their query. Most queries I have seen use `default` in pretty localized places, and not as "catch alls" over a long sequence of terms. So the risk of unintentionally catching and ignoring a numeric problem is probably not very high.

I do think that the question of which errors to catch in `default` matters. We will eventually allow providing a list of error types to `default`, but - unless it turns out by then that our decision has been substantially wrong - we will likely try to keep the default behavior that we decide on now.

Anyway, I think the easier solution of having a generic `NumericError` error type for whenever `Inf` or `NaN` pop up, and catching that in `default` is good.

**Comment by coffeemug** [comment]
:+1:

**Comment by danielmewes** [comment]
We haven't fully resolved the question about the name of the new error type yet. Some options that were mentioned:
* `ReqlArithmeticError` (similar to Python)
* `ReqlFloatError`
* `ReqlNumericError`
* `ReqlMathError`

I personally like `ReqlNumericError`.

**Comment by mlucy** [comment]
`NumericError` works for me.

**Comment by coffeemug** [comment]
:+1:

**Comment by danielmewes** [comment]
Marking settled as:
* adding a new error type `ReqlNumericError` to be thrown whenever we encounter an `Inf` or `NaN` value
* making `default` catch `ReqlNumericError` in addition to non existence errors

**Comment by marshall007** [comment]
Just realized my proposal in [#5661 (comment)](https://github.com/rethinkdb/rethinkdb/issues/5661#issuecomment-209595470) is pretty much identical to the long-term solution brought up in this issue. I think extending `default` to accept error types makes more sense than adding a new term as I was suggesting. That said, I'd prefer specifying the error type as the first argument rather than an optarg as well as support for passing actual `Error` types in addition to strings. That way it reads more like a traditional `try/catch`.

```js
expr.default('Math', 'NonExistence', 0)

expr.default(r.Error.ReqlMathError, r.Error.ReqlNonExistenceError, 0)

expr.default(r.Error.ReqlMathError, 0)
    .default(r.Error.ReqlNonExistenceError, 0)
```
